net/portmapper/upnp.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package portmapper
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"inet.af/netaddr"
+	"tailscale.com/tempfork/upnp/dcps/internetgateway2"
+)
+
+type upnpMapping struct {
+	gw       netaddr.IP
+	external netaddr.IPPort
+	internal netaddr.IPPort
+	useUntil time.Time
+	client   upnpClient
+}
+
+func (u *upnpMapping) isCurrent() bool                { return u.useUntil.After(time.Now()) }
+func (u *upnpMapping) validUntil() time.Time          { return u.useUntil }
+func (u *upnpMapping) externalIPPort() netaddr.IPPort { return u.external }
+func (u *upnpMapping) release() {
+	u.client.DeletePortMapping(context.Background(), "", u.external.Port(), "udp")
+}
+
+type upnpClient interface {
+	// http://upnp.org/specs/gw/UPnP-gw-WANIPConnection-v2-Service.pdf
+	// Implicitly assume that the calls for all these are uniform, which might be a dangerous
+	// assumption.
+	AddPortMapping(
+		ctx context.Context,
+		newRemoteHost string,
+		newExternalPort uint16,
+		newProtocol string,
+		newInternalPort uint16,
+		newInternalClient string,
+		newEnabled bool,
+		newPortMappingDescription string,
+		newLeaseDuration uint32,
+	) (err error)
+
+	DeletePortMapping(ctx context.Context, NewRemoteHost string, NewExternalPort uint16, NewProtocol string) error
+	GetStatusInfo(ctx context.Context) (status string, lastErr string, uptime uint32, err error)
+	GetExternalIPAddress(ctx context.Context) (externalIPAddress string, err error)
+
+	RequestTermination(ctx context.Context) error
+	RequestConnection(ctx context.Context) error
+}
+
+func AddAnyPortMapping(
```

Comment:
too many parameters. Also, I thought you were going to make this a method some stateful type?